Client/Frontend/Browser/PasswordHelper.swift

```diff
@@ -171,4 +172,95 @@ class PasswordHelper: BrowserHelper {
             })
         })
     }
+
+    func handleAuthRequest(viewController: UIViewController, challenge: NSURLAuthenticationChallenge, completion: (password: Password?) -> Void) {
+        // If there have already been too many login attempts, we'll just fail.
+        if challenge.previousFailureCount > PasswordHelper.MaxAuthenticationAttempts {
+            completion(password: nil)
+            return
+        }
+
+        var credential = challenge.proposedCredential
+
+        // If we were passed an initial set of credentials from iOS, try and use them.
+        if let proposed = credential {
+            if let user = proposed.user {
+                if !user.isEmpty {
+                    if challenge.previousFailureCount == 0 {
+                        println("Use first guess")
+                        completion(password: Password(credential: credential!, protectionSpace: challenge.protectionSpace))
+                        return
+                    }
+                } else {
+                    credential = nil
+                }
+            } else {
+                credential = nil
+            }
+        }
+
+        // If we have some credentials, we'll show a prompt with them.
+        if let credential = credential {
+            println("Use guess")
+            let password = Password(credential: credential, protectionSpace: challenge.protectionSpace)
+            promptForUsernamePassword(viewController, password: password, completion: completion)
+        } else {
+            // Otherwise, try to look some up
+            let options = QueryOptions(filter: challenge.protectionSpace.host, filterType: .None, sort: .None)
+            profile.passwords.get(options, complete: { (cursor) -> Void in
+                var password = cursor[0] as? Password
+                println("Found \(password?.username)")
+                if password == nil {
+                    password = Password(credential: nil, protectionSpace: challenge.protectionSpace)
+                }
+                self.promptForUsernamePassword(viewController, password: password!, completion: completion)
+            })
+        }
+    }
+
+    private func promptForUsernamePassword(viewController: UIViewController, password: Password, completion: (password: Password?) -> Void) {
+
+        // Show an alert controller (prompt) to get the username and password.
+        let alert = UIAlertController(title: NSLocalizedString("Authentication required", comment: "Authentication prompt title"),
+            message: NSLocalizedString(NSString(format: "A username and password are being requested by %@. The site says: %@",
```

Comment:
I don't think this is right. You're doing the substitution *before* handing the string to `NSLocalizedString`, so what is the actual string that will be exported?

I could be wrong, but I think you'll need to have a separate `NSLocalizedString` for the "(unknown)" strings. Then you would do `NSString(format: NSLocalizedString(...))` instead of the other way around.